server/datastore/mysql/microsoft_mdm.go

```diff
@@ -135,210 +128,214 @@ func (ds *Datastore) MDMWindowsDeleteEnrolledDeviceWithDeviceID(ctx context.Cont
+VALUES %s
+ON DUPLICATE KEY UPDATE
+    raw_result = COALESCE(VALUES(raw_result), raw_result),
+    status_code = COALESCE(VALUES(status_code), status_code)
+	`
+
+	enrollment, err := ds.MDMWindowsGetEnrolledDeviceWithDeviceID(ctx, deviceID)
 	if err != nil {
-		return ctxerr.Wrap(ctx, err, "updating windows command rx_cmd_result")
+		return ctxerr.Wrap(ctx, err, "getting enrollment with device ID")
 	}
 
-	return nil
-}
+	return ds.withRetryTxx(ctx, func(tx sqlx.ExtContext) error {
+		// grab all the incoming UUIDs
+		var cmdUUIDs []string
+		uuidsToStatus := make(map[string]fleet.SyncMLCmd)
+		uuidsToResults := make(map[string]fleet.SyncMLCmd)
+		for _, protoOp := range fullResponse.GetOrderedCmds() {
+			// results and status should contain a command they're
+			// referencing
+			cmdRef := protoOp.Cmd.CmdRef
+			if !protoOp.Cmd.ShouldBeTracked(protoOp.Verb) || cmdRef == nil {
+				continue
+			}
 
-// MDMWindowsListCommands retrieves all commands for a given device ID from the windows_mdm_commands table
-func (ds *Datastore) MDMWindowsListCommands(ctx context.Context, deviceID string) ([]*fleet.MDMWindowsCommand, error) {
-	var commands []*fleet.MDMWindowsCommand
+			switch protoOp.Verb {
+			case fleet.CmdStatus:
+				uuidsToStatus[*cmdRef] = protoOp.Cmd
+				cmdUUIDs = append(cmdUUIDs, *cmdRef)
+			case fleet.CmdResults:
+				uuidsToResults[*cmdRef] = protoOp.Cmd
+				cmdUUIDs = append(cmdUUIDs, *cmdRef)
+			}
+		}
 
-	query := `
-        SELECT
-            command_uuid,
-            device_id,
-            session_id,
-            message_id,
-            command_id,
-            cmd_verb,
-            setting_uri,
-            setting_value,
-            system_origin,
-            rx_error_code,
-            rx_cmd_result,
-            created_at,
-            updated_at
-        FROM
-            old_windows_mdm_commands
-        WHERE
-            device_id = ?
-    `
+		// no relevant commands to tracks is a noop
+		if len(cmdUUIDs) == 0 {
+			return nil
+		}
 
-	if err := sqlx.SelectContext(ctx, ds.reader(ctx), &commands, query, deviceID); err != nil {
-		return nil, ctxerr.Wrap(ctx, err, "get Windows MDM commands by device id")
-	}
+		// store the full response
+		sqlResult, err := tx.ExecContext(ctx, saveFullRespStmt, enrollment.ID, fullResponse.Raw)
+		if err != nil {
+			return ctxerr.Wrap(ctx, err, "saving full response")
+		}
+		responseID, _ := sqlResult.LastInsertId()
 
-	return commands, nil
+		// find commands we sent that match the UUID responses we've got
+		stmt, params, err := sqlx.In(findCommandsStmt, cmdUUIDs)
+		if err != nil {
+			return ctxerr.Wrap(ctx, err, "building IN to search matching commands")
+		}
+		var matchingUUIDs []string
+		err = sqlx.SelectContext(ctx, tx, &matchingUUIDs, stmt, params...)
+		if err != nil {
+			return ctxerr.Wrap(ctx, err, "selecting matching commands")
+		}
+
+		// TODO(roberto): no matching commands is a noop? should we
```

Comment:
Yeah I think that's ok. I think we'd want to explicitly store only those that we care about (previously stored as pending commands).